sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java

```diff
@@ -0,0 +1,314 @@
+    }
+
+    @Override
+    public List<? extends Coder<?>> getCoderArguments() {
+      return Arrays.<Coder<?>>asList(valueCoder);
+    }
+
+    public Coder<T> getValueCoder() {
+      return valueCoder;
+    }
+
+    @Override
+    public TypeDescriptor<TimestampedValue<T>> getEncodedTypeDescriptor() {
+      return new TypeDescriptor<TimestampedValue<T>>() {}.where(
+          new TypeParameter<T>() {}, valueCoder.getEncodedTypeDescriptor());
+    }
+
+    @Override
+    public List<? extends Coder<?>> getComponents() {
+      return Collections.singletonList(valueCoder);
+    }
+  }
+
+  public OrderedListUserState(
+      Cache<?, ?> cache,
+      BeamFnStateClient beamFnStateClient,
+      String instructionId,
+      StateKey stateKey,
+      Coder<T> valueCoder) {
+    checkArgument(
+        stateKey.hasOrderedListUserState(),
+        "Expected OrderedListUserState StateKey but received %s.",
+        stateKey);
+    this.beamFnStateClient = beamFnStateClient;
+    this.timestampedValueCoder = TimestampedValueCoder.of(valueCoder);
+    this.request =
+        StateRequest.newBuilder().setInstructionId(instructionId).setStateKey(stateKey).build();
+  }
+
+  public void add(TimestampedValue<T> value) {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+    Instant timestamp = value.getTimestamp();
+    pendingAdds.putIfAbsent(timestamp, new ArrayList<>());
+    pendingAdds.get(timestamp).add(value.getValue());
+  }
+
+  public Iterable<TimestampedValue<T>> readRange(Instant minTimestamp, Instant limitTimestamp) {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    // Store pendingAdds whose sort key is in the query range and values are truncated by the
+    // current size. The values (collections) of pendingAdds are kept, so that they will still be
+    // accessible in pre-existing iterables even after:
+    //   (1) a sort key is added to or removed from pendingAdds, or
+    //   (2) a new value is added to an existing sort key
+    ArrayList<PrefetchableIterable<TimestampedValue<T>>> pendingAddsInRange = new ArrayList<>();
+    for (Entry<Instant, Collection<T>> kv :
+        pendingAdds.subMap(minTimestamp, limitTimestamp).entrySet()) {
+      pendingAddsInRange.add(
+          PrefetchableIterables.limit(
+              Iterables.transform(kv.getValue(), (v) -> TimestampedValue.of(v, kv.getKey())),
+              kv.getValue().size()));
+    }
+    Iterable<TimestampedValue<T>> valuesInRange = Iterables.concat(pendingAddsInRange);
+
+    if (!isCleared) {
+      StateRequest.Builder getRequestBuilder = this.request.toBuilder();
+      getRequestBuilder
+          .getStateKeyBuilder()
+          .getOrderedListUserStateBuilder()
+          .getRangeBuilder()
+          .setStart(minTimestamp.getMillis())
```

Comment:
I think we should have a discussion about if the sort key here is really a timestamp or not (and whether we should be using TimestampedValue). This leads to weird things like using BoundedWindow.min/max and having to convert Instants to ints (in millis). I thought the sort key could be any int64 and doesn't have to be in the timestamp range.

Reply:
I think here we assume int64 and instant are interchangeable. (Original Design Doc Reference: https://docs.google.com/document/d/1U77sAvE6Iy9XsVruRYHxPdFji7nqS6HPi1XU8fhyrxs/edit?disco=AAAAHwml1Yg . In such a case, the min and max values should be Long.MIN_VALUE and Long.MAX_VALUE. 

The ordered list state is designed around TimestampedValue, but it can be used on int64 keys as well.

Reply:
Fixed.